source/geod24/concurrency.d

```diff
@@ -925,35 +925,45 @@ protected:
     static class InfoFiber : Fiber
     {
         ThreadInfo info;
+        FiberCondition cond;
 
         this(void delegate() op, size_t sz = 512 * 1024) nothrow
         {
+            cond = null;
             super(op, sz);
         }
     }
 
     protected class FiberCondition
     {
+        MonoTime limit = MonoTime.init;
+
+        private void registerToInfoFiber() nothrow
+        {
+            /*
+            *  Register the FiberCondition to InfoFiber so that FiberScheduler can check if condition is met
+            */
+            auto infoF = cast(InfoFiber) Fiber.getThis();
+            assert(infoF !is null, "This Fiber does not belong to FiberScheduler");
+            assert(infoF.cond is null, "This Fiber already has a registered FiberCondition");
+            infoF.cond = this;
+        }
+
         void wait() nothrow
         {
+            registerToInfoFiber();
+            FiberScheduler.yield();
             scope (exit) notified = false;
-
-            while (!notified)
-                FiberScheduler.yield();
         }
 
         bool wait(Duration period) nothrow
         {
-            import core.time : MonoTime;
+            registerToInfoFiber();
+            limit = MonoTime.currTime + period;
+            FiberScheduler.yield();
+            limit = MonoTime.init;
```

Comment:
Why is resetting the limit necessary? The `FiberCondition` is never re-used as far as I know. And it might hinder debugging if the fields are reset after use.

Reply:
`notified` gets reset after `wait` , I simply followed the same pattern. If we will rely on them not being reused we should have an assertion for it I guess.

Reply:
@Geod24 , what is your take on this?

Reply:
I think those variables should be reset when entering the function. In another review comment you mentioned that we could keep the possibility of multiple fibers waiting on the same condition, but that would not work if we reset `notified` at the end of the function here.

Reply:
Oh yes, that would not work. There is another issue with resetting them at the start though. Image creating a condition and passing a reference of it to another thread. If condition gets signaled before you wait on it, you will miss that event and may even end up waiting indefinitely. I am not sure if this can be solved with additional locking, but I doubt it.

Reply:
Alright, let's keep it as is for now